Added postprocessor to compute change in a postprocessor over a timestep

joshuahansel opened the pull request:
closes #8657

joshuahansel committed:
Added postprocessor to compute change in a postprocessor over a timestep

closes #8657

joshuahansel commented:
@andrsd

andrsd commented on test/tests/postprocessors/change_over_timestep/change_over_timestep.i:
```diff
@@ -0,0 +1,77 @@
+
+[Problem]
+  type = FEProblem
+[]
```
This should be the default, so you can remove it.

andrsd commented on test/tests/postprocessors/change_over_timestep/change_over_timestep.i:
```diff
@@ -0,0 +1,77 @@
+
+[Executioner]
+  type = Transient
+  scheme = implicit-euler
```
`implicit-euler` is the default, so this can go...

andrsd commented on framework/src/postprocessors/ChangeOverTimestepPostprocessor.C:
```diff
@@ -0,0 +1,46 @@
+Real
+ChangeOverTimestepPostprocessor::getValue()
+{
+  return _pps_value - _pps_value_old;
```
We typically compute the PPS value in `execute()` and `getValue` acts like a getter. Typically it has some parallel communication in it as well. I guess this will work in parallel, since `_pps_value` and `_pps_value_old` are the same on all procs.

In case parallel test fails, this might be a source of problems.

joshuahansel replied:
I followed the example of `DifferencePostprocessor` for this; it does what mine is doing - it just computes the value in `getValue` and does nothing for `threadJoin`. The comment in `threadJoin` says "nothing to do here, general PPS do not run threaded", which is probably what you expected and makes sense.

andrsd replied:
`threadJoin` is for threading. Another type of communication is MPI-related. That's what we typically do in `getValue()`.  This is most likely fine and will work.

mangerij commented:
I should mention that we do have `PercentChangePostprocessor` which does something similar. Maybe we can just put some bool option to either compute the percent change or the actual change using different functions instead of adding another file.

joshuahansel commented:
@mangerij I am ok with that suggestion. I think that I will incorporate `PercentChangePostprocessor` into the new postprocessor (since the new one has a name that generalizes to both cases) and deprecate `PercentChangePostprocessor`. Furthermore, the name `PercentChangePostprocessor` needed renaming anyway since it's not computing *percent* change but instead, *relative* change.

joshuahansel committed:
Incorporated PercentChangePostprocessor into ChangeOverTimestepPostprocessor

As suggested, PercentChangePostprocessor is similar enough to ChangeOverTimestepPostprocessor
that they could be combined. Since ChangeOverTimestepPostprocessor had a more general name,
PercentChangePostprocessor was incorporated into ChangeOverTimestepPostprocessor and then
deprecated.

refs #8657

andrsd commented on framework/src/postprocessors/PercentChangePostprocessor.C:
```diff
@@ -31,6 +31,9 @@ PercentChangePostprocessor::PercentChangePostprocessor(const InputParameters & p
 {
+  mooseDeprecated("PercentChangePostprocessor is deprecated: instead, ",
+    "please use ChangeOverTimestepPostprocessor using the parameter ",
+    "'compute_relative_change' set to 'true'");
```
Move this into a c-tor, `initialize()` can be called as much as once per residual evaluation and this would flood people's terminals.  That would make it annoying enough so they would update their codes, but still... ;-)
